tools/nnyeah/nnyeah/MethodTransformations.cs

```diff
@@ -0,0 +1,832 @@
+using System;
+using System.Collections.Generic;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+namespace nnyeah {
+	public class MethodTransformations {
+		static List<Transformation> allTransforms;
+		static Dictionary<string, Transformation> transformTable;
+
+		public Dictionary<string, Transformation> GetTransforms (ModuleDefinition module, Func<List<bool>, CustomAttribute> attrBuilder)
+		{
+			if (allTransforms != null) {
+				return transformTable;
+			}
+
+			// there are two types of transforms here: ones that can be made statically
+			// and ones that need more in the way of state and intermediate variables.
+
+			var singleBool = new List<bool> () { true };
+
+			allTransforms = new List<Transformation> (transforms.Length);
+			allTransforms.AddRange (transforms);
+
+			// nint
+			var mref = new MethodReference ("CompareTo", module.TypeSystem.Int32, module.TypeSystem.IntPtr);
+			var parm = new ParameterDefinition (module.TypeSystem.IntPtr);
+			parm.CustomAttributes.Add (attrBuilder (singleBool));
+			mref.Parameters.Add (parm);
+			allTransforms.Add (new Transformation ("System.Int32 System.nint::CompareTo(System.nint)",
+				Instruction.Create (OpCodes.Call, mref)));
+
+			mref = new MethodReference ("CompareTo", module.TypeSystem.Int32, module.TypeSystem.IntPtr);
```

Comment:
What is the diff between this MethodReference:
```csharp
			mref = new MethodReference ("CompareTo", module.TypeSystem.Int32, module.TypeSystem.IntPtr);
```
and the one in line 26?
```csharp
			var mref = new MethodReference ("CompareTo", module.TypeSystem.Int32, module.TypeSystem.IntPtr);
```
Do they have to be two difference intances?

Reply:
AFAIK, they do have to be two different instances.
The general philosophy of this code is that each of the transformations is a singleton and we should have as few shared references as possible. This is to prevent errors in the future where what looks like a localized change affects a bunch of other things.

Reply:
Hmm would a value type like a record record structs help?

Reply:
I believe the MethodReference are cecil types, so outside of wrapping them I don't think he can change that bit.

I wonder if this would look better as each complex transform being it's own little function:

```
static Transformation CompareToTransform ()
{
    var mref = new MethodReference ("CompareTo", module.TypeSystem.Int32, module.TypeSystem.IntPtr);
    var parm = new ParameterDefinition (module.TypeSystem.IntPtr);
    parm.CustomAttributes.Add (attrBuilder (singleBool));
    mref.Parameters.Add (parm);
    return new Transformation ("System.Int32 System.nint::CompareTo(System.nint)", Instruction.Create (OpCodes.Call, mref));
```
```
    allTransforms.Add (CompareToTransform ();
    allTransforms.Add (FooTransform ();
    allTransforms.Add (BarTransform ();
```